cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,564 @@
+#ifndef CAJITA_SPARSE_INDEXSPACE_HPP
+#define CAJITA_SPARSE_INDEXSPACE_HPP
+
+#include <Kokkos_Atomic.hpp>
+#include <Kokkos_Core.hpp>
+#include <Kokkos_UnorderedMap.hpp>
+
+#ifdef __CUDACC__
+#define CUDA_HOSTDEV __host__ __device__
+#else
+#define CUDA_HOSTDEV
+#endif
+
+#include <array>
+#include <string>
+
+namespace Cajita
+{
+
+enum class HashTypes : unsigned char
+{
+    Naive = 0,
+    Morton = 1
+};
+
+//---------------------------------------------------------------------------//
+// bit operations for Morton Code computing
+template <typename Integer>
+CUDA_HOSTDEV constexpr Integer bit_length( Integer N ) noexcept
+{
+    if ( N )
+        return bit_length( N >> 1 ) + static_cast<Integer>( 1 );
+    else
+        return 0;
+}
+
+template <typename Integer>
+KOKKOS_FORCEINLINE_FUNCTION constexpr Integer bit_count( Integer N ) noexcept
+{
+    return bit_length( N - 1 );
+}
+
+template <typename Integer>
+CUDA_HOSTDEV constexpr Integer
+binary_reverse( Integer data, char loc = sizeof( Integer ) * 8 - 1 )
+{
+    if ( data == 0 )
+        return 0;
+    return ( ( data & 1 ) << loc ) | binary_reverse( data >> 1, loc - 1 );
+}
+
+template <typename Integer>
+KOKKOS_FORCEINLINE_FUNCTION constexpr unsigned
+count_leading_zeros( Integer data )
+{
+    unsigned res{0};
+    data = binary_reverse( data );
+    if ( data == 0 )
+        return sizeof( Integer ) * 8;
+    while ( ( data & 1 ) == 0 )
+        res++, data >>= 1;
+    return res;
+}
+
+KOKKOS_FORCEINLINE_FUNCTION
+constexpr int bit_pack( const uint64_t mask, const uint64_t data )
+{
+    uint64_t slresult = 0;
+    uint64_t &ulresult{slresult};
+    uint64_t uldata = data;
+    int count = 0;
+    ulresult = 0;
+
+    uint64_t rmask = binary_reverse( mask );
```

Comment:
Longer functions like this should have comments internally as well to describe what you are doing.